Add x-checker-data for kolibri-explore-plugin module

starnight opened the pull request:
Fixes: https://github.com/endlessm/endless-key-flatpak/issues/45

dylanmccall commented:
Thanks for taking care of this, @StarNight! I pushed a small change which uses the JSON checker against https://api.github.com/repos/endlessm/kolibri-explore-plugin/releases/latest to get the newest download files. That checker is documented over here: https://github.com/flathub/flatpak-external-data-checker#json-checker.

That works nicely for me, now, but I'm wondering what you all think about getting this to fetch the whl from github as well? We do have it there, and I feel like fetching the whl from pypi and everything else from gihub introduces the possibility of things getting out of sync.

starnight commented:
> It works nicely for me with that change, but I'm wondering what you all think about getting this to fetch the whl from github as well? We do have it there, and I feel like fetching the whl from pypi and everything else from github introduces the possibility of things getting out of sync.

Good question!  However, `kolibri_explore_plugin-7.1.0-py3-none-any.whl` includes the version string in the release page.  Besides, the wheel, apps-bundle and loading-screen are released in the same [release job](https://github.com/endlessm/kolibri-explore-plugin/blob/910e8b3e108ba8f7a3ed9e4a12cda557615ed1d2/.github/workflows/release.yml#L58).  It might be okay.

starnight commented:
Ok!  I found a way to deal the version string in `kolibri_explore_plugin-7.1.0-py3-none-any.whl`.  Updated the commit.

Besides, I also added `x-checker-data` into other modules, too.  But, skip **kolibri** module.  Because, Endless Key uses kolibri's pre-release right now.

dbnicholson commented:
> It works nicely for me with that change, but I'm wondering what you all think about getting this to fetch the whl from github as well? We do have it there, and I feel like fetching the whl from pypi and everything else from github introduces the possibility of things getting out of sync.

I've definitely thought about it a lot, but I think it's probably fine. Certainly they're the same file so long as no one is doing a manual upload to PyPI.

dbnicholson commented:
One thing I'd really like to see here is version constraints. Should they happen, we would not want to automatically upgrade to kolibri-explore-plugin 8.y or kolibri 0.17.z.

dbnicholson reviewed (changes requested):
I'm certainly in agreement with using this, but I'm not as excited about blindly updating major versions of transitive dependencies.

starnight commented on build-aux/flatpak/modules/iproute2.json:
```diff
@@ -8,9 +8,13 @@
+            "x-checker-data" : {
+                "type" : "git",
+                "tag-pattern" : "^v([\\d.]+)$"
+            }
```
I tested with iproute2 6.5.0.  Endless Key works as well.
Kolibri depends on [ifcfg](https://github.com/learningequality/kolibri/blob/v0.16.0-beta5/requirements/base.txt#L29) and the ifcfg package uses `ip` command from iproute2.

starnight commented on build-aux/flatpak/modules/python3-kolibri-app-desktop-xdg-plugin.json:
```diff
@@ -8,12 +8,21 @@
+            "sha256": "dde3f3ed8d00c72631bc19cbfff8ad3b6215062a5eed402381ad365f82f0c18c",
+            "x-checker-data": {
+                "type": "pypi",
+                "name": "Pillow"
```
Notice the restriction `Pillow>=8.2.0,<9.0.0` in [kolibri-app-desktop-xdg-plugin](https://github.com/endlessm/kolibri-app-desktop-xdg-plugin/blob/v1.1.4/setup.py#L28).

starnight replied:
Updated the commit.

starnight commented on build-aux/flatpak/modules/python3-virtualenv-api.json:
```diff
@@ -8,7 +8,12 @@
+            "sha256": "8b74bedcbbbaca38ff6d7491d76f2b06b3592611af620f8426e82dddb04a5ced",
+            "x-checker-data": {
+                "type": "pypi",
+                "name": "six",
```
I have tested without `modules/python3-virtualenv-api.json`.  Endless Key builds and works as usual.  I am going to drop `modules/python3-virtualenv-api.json`.  Please blame me if it is needed.

starnight commented:
Also, add `x-checker-data` into kolibri module.

dbnicholson commented on build-aux/flatpak/modules/python3-kolibri-explore-plugin.json:
```diff
@@ -10,21 +10,42 @@
+                "url": "https://api.github.com/repos/endlessm/kolibri-explore-plugin/releases/latest",
+                "version-query": ".tag_name",
+                "url-query": ".assets[] | select(.name==\"apps-bundle.zip\") | .browser_download_url",
+                "versions": {"<": "8.0.0"}
```
As it turns out, the JSON checker doesn't support `versions`, so we'll just have to do this with `jq` filters.

dbnicholson reviewed (approved):
This is good enough to start even though some of the version constraints won't work. I rebased to address the conflict and added the `v` stripping for the apps-bundle and loading-screen versions.

starnight commented on build-aux/flatpak/modules/python3-kolibri.json:
```diff
@@ -9,7 +9,14 @@
+                "type": "json",
+                "url": "https://api.github.com/repos/learningequality/kolibri/releases",
+                "version-query": "first | .tag_name | sub(\"^[vV]\"; \"\") | sub(\"-beta\"; \"b\")",
+                "url-query": "first | .assets[] | select(.name==\"kolibri-\" + $version + \"-py2.py3-none-any.whl\") | .browser_download_url",
```
The `/latest` only shows the latest normal release.  But, we are using Kolibri's **pre-release**, which will not be shown by `/latest`.  So, we can only pick the first release by ourselves.
